### PR TITLE
fix: Frameless window shows frame while opening

### LINKED ISSUE #30760: [Bug]: Frameless window displays frame while opening
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

14.0.0-beta.25

### What operating system are you using?

Windows

### Operating System Version

Windows 10 21H1

### What arch are you using?

x64

### Last Known Working Electron version

14.0.0-beta.3

### Expected Behavior

Opening a frameless window and immediately revealing it (instead of waiting for `ready-to-show` event) shows no content inside of the window.

### Actual Behavior

Opening a frameless window and immediately revealing it (instead of waiting for `ready-to-show` event) shows an older style window frame with no content inside.

| 14.0.0-beta.3 | 14.0.0-beta.5 |
| --- | --- |
| image |  image |

https://user-images.githubusercontent.com/1656324/131413659-18c2ea6c-d1c4-489f-8182-ae19cba51bcb.mp4

### Testcase Gist URL

https://gist.github.com/4256f388d41ef0d974645a11ab72b1fe

### Additional Information

This may not be a bug, but I felt that displaying the frame while the window is opening was maybe not intentional.

### PR DESCRIPTION
**Pull request by rzhao271**
#### Description of Change

Fixes #30760

The frame was showing because https://github.com/electron/electron/pull/29594 removed the `CanResize` accessor on the Electron side, which made it so that `WidgetDelegate::CanResize` started initially returning `false`. I have to do more investigation to see why `CanResize` returning `false` early on leads to the frame showing up. I have confirmed that the issue doesn't have to do with `SetCanResize` making the widget emit `OnSizeConstraintsChanged`.

Note that this PR only fixes the issue frameless windows that are resizable. It does not fix the issue for frameless windows that are not resizable (see https://github.com/electron/electron/issues/30024#issuecomment-1212056696).

CC @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed an issue where frameless resizable windows showed Window 7 style frames while opening.

### CODE REVIEW

File: patches/chromium/feat_add_set_can_resize_mutator.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+diff --git a/ui/views/widget/widget_delegate.cc b/ui/views/widget/widget_delegate.cc
+index 6d58e740b5693c0f70eecce88593851f43c5d020..41a233a3c7a5bbf88fa13bd43ec607fda53cd7de 100644
+--- a/ui/views/widget/widget_delegate.cc
++++ b/ui/views/widget/widget_delegate.cc
+@@ -299,6 +299,10 @@ bool WidgetDelegate::ShouldDescendIntoChildForEventHandling(
+   return true;
+ }
+ 
++void WidgetDelegate::set_can_resize(bool can_resize) {
```
Comment (deepak1556) [inline patches/chromium/feat_add_set_can_resize_mutator.patch:21]: setters can be inlined to header file.

File: shell/browser/native_window_views.cc [hunk 2]
```diff
@@ -281,7 +281,13 @@ NativeWindowViews::NativeWindowViews(const gin_helper::Dictionary& options,
       new ElectronDesktopWindowTreeHostLinux(this, native_widget);
 #endif
 
+  // We want to set the can_resize param before initializing
+  // the widget, but we want to do it without emitting any events.
```
Comment (deepak1556) [inline shell/browser/native_window_views.cc:285]: Please explain why we want `can_resize` to be true before initializing the widget.

**Review by deepak1556 — CHANGES_REQUESTED** [review]
For context on this change please read up https://github.com/microsoft/vscode/issues/146683#issuecomment-1212284906

What we are doing in this PR is restoring the behavior of framless resizable windows to have the same styles as before https://github.com/electron/electron/commit/172ac2501330c64c2a33c1004e05eb23f9658903 . Basically the presence of `WS_THICKFRAME` before the widget creation has an impact on the NC messages received and hiding the resizable frame. We still don't know why this happens and we are following up with DWM team on windows to understand better so that we can also fix this for the non-resizable frameless scenario https://github.com/electron/electron/issues/30024.

File: shell/browser/native_window_views.cc (hunk 2 above)
Comment (rzhao271) [reply shell/browser/native_window_views.cc:285]: Actually, I never tried setting it to true directly. I will have to confirm whether that could actually temporarily resolve the non-resizable scenario as well.

Comment (rzhao271) [reply shell/browser/native_window_views.cc]: If we want the issue to be resolved for non-resizable windows as well, then it turns out we have to move the `SetCanResize` call to after the `SetWindowLong` call in this file. Worse would be to patch Chromium to add in `WS_THICKFRAME` styles. I clarified the comments.

File: shell/browser/native_window_views.cc [hunk 3]
```diff
@@ -281,7 +281,23 @@ NativeWindowViews::NativeWindowViews(const gin_helper::Dictionary& options,
...
+  // When resizable_ is true, this causes the WS_THICKFRAME style
+  // to be passed into CreateWindowEx and SetWindowLong calls in
+  // WindowImpl::Init and HwndMessageHandler::SizeConstraintsChanged
+  // respectively. As a result, the Windows 7 frame doesn't show,
+  // but it isn't clear why this is the case.
+  // When resizable_ is false, WS_THICKFRAME is not passed into either
+  // call, so the Windows 7 frame still shows. One workaround would be
+  // to call set_can_resize(true) here, and then move the
+  // SetCanResize(resizable_) call after the SetWindowLong call
+  // around line 365, but that's a much larger change.
```
Comment (zcbenz) [inline shell/browser/native_window_views.cc:294]: Can you add a link to the vscode issue in the comment?

**Review by zcbenz — APPROVED** [review]

**Review by deepak1556 — APPROVED** [review]

**Review by MarshallOfSound — APPROVED** [review]

**Pull request merged**